src/Block/ChildrenPagesBlockService.php

```diff
@@ -178,15 +176,30 @@ public function load(BlockInterface $block): void
 
         $pageId = $block->getSetting('pageId', null);
 
-        if (null !== $pageId && !$pageId instanceof PageInterface) {
-            $cmsManager = $this->cmsManagerSelector->retrieve();
+        if (!$pageId instanceof PageInterface) {
+            $block->setSetting('pageId', $this->pageAdmin->getObject($pageId));
+        }
+    }
 
-            $page = $block->getPage();
-            $site = null !== $page ? $page->getSite() : null;
+    /**
+     * @param AdminFormMapper<object> $form
+     */
+    private function getPageSelectorBuilder(AdminFormMapper $form, PageBlockInterface $block): FormBuilderInterface
+    {
+        $page = $block->getPage();
 
-            if (null !== $site) {
-                $block->setSetting('pageId', $cmsManager->getPage($site, $pageId));
-            }
-        }
+        $formBuilder = $form->getFormBuilder()->create('pageId', PageSelectorType::class, [
+            'model_manager' => $this->pageAdmin->getModelManager(),
+            'class' => $this->pageAdmin->getClass(),
+            'site' => null !== $page ? $page->getSite() : null,
+            'required' => false,
+            'label' => 'form.label_page',
+        ]);
+        $formBuilder->addModelTransformer(new CallbackTransformer(
```

Comment:
The problem @VincentLanglet is that when you create the field for the 'pageId', pageId is not a mapped field on any entity, and when it gets to doctrine to store the information, it gets in the form of a Page (PageInterface), doctrie does not understand that I want to store only the id (It does for mapped fields tho) and with the model transformer we solve the issue.

Previously on 3.x we had some prePersist and preUpdate hooks that essentially were doing the same transformation, this solution looks cleaner to me.

Reply:
Seems better to have a standalone form indeed rather than a prepist/preupdate fix.